crates/holochain/tests/inline_zome_spec.rs

```diff
@@ -174,3 +205,76 @@ async fn invalid_cell() -> anyhow::Result<()> {
 
     Ok(())
 }
+
+/// Simple zome which contains a validation rule which can fail
+fn simple_validation_zome() -> InlineZome {
+    let entry_def = EntryDef::default_with_id("string");
+
+    InlineZome::new_unique(vec![entry_def.clone()])
+        .callback("create", move |api, s: AppString| {
+            let entry_def_id: EntryDefId = entry_def.id.clone();
+            let entry = Entry::app(s.try_into().unwrap()).unwrap();
+            let hash = api.create((entry_def_id, entry))?;
+            Ok(hash)
+        })
+        .callback("read", |api, hash: HeaderHash| {
+            api.get((hash.into(), GetOptions::default()))
+                .map_err(Into::into)
+        })
+        .callback("validate_create_entry", |_api, data: ValidateData| {
+            let s: AppString = data.element.entry().to_app_option().unwrap().unwrap();
+            if &s.0 == "" {
+                Ok(ValidateResult::Invalid("No empty strings allowed".into()))
+            } else {
+                Ok(ValidateResult::Valid)
+            }
+        })
+}
+
+#[tokio::test(threaded_scheduler)]
+async fn simple_validation() -> anyhow::Result<()> {
+    let (dna_file, _) =
+        CoolDnaFile::unique_from_inline_zome("zome", simple_validation_zome()).await?;
+    let mut conductor = CoolConductor::from_standard_config().await;
+    let (alice, bobbo) = CoolAgents::two(conductor.keystore()).await;
+    let apps = conductor
+        .setup_app_for_agents("app", &[alice.clone(), bobbo.clone()], &[dna_file])
+        .await;
+    let ((alice,), (bobbo,)) = apps.into_tuples();
+
+    let alice = alice.zome("zome");
+    let _bobbo = bobbo.zome("zome");
+
+    // This call passes validation
+    let h1: HeaderHash = conductor
+        .call(&alice, "create", AppString::new("A"))
+        .await?;
+    let e1: MaybeElement = conductor.call(&alice, "read", &h1).await?;
+    let s1: AppString = e1.0.unwrap().entry().to_app_option().unwrap().unwrap();
+    assert_eq!(s1, AppString::new("A"));
+
+    // This call fails validation, and so results in an error
+    let err: ConductorApiResult<HeaderHash> =
+        conductor.call(&alice, "create", AppString::new("")).await;
+
+    // This is kind of ridiculous, but we can't use assert_matches! because
+    // there is a Box in the mix.
```

Comment:
There's a fix for this coming https://doc.rust-lang.org/beta/unstable-book/language-features/box-patterns.html